src/roaring.c

```diff
@@ -2379,51 +2307,50 @@ roaring_bitmap_t *roaring_bitmap_lazy_xor(const roaring_bitmap_t *x1,
     return answer;
 }
 
-void roaring_bitmap_lazy_xor_inplace(roaring_bitmap_t *x1,
-                                     const roaring_bitmap_t *x2) {
+
+// Common code for "xor_inplace" and "xor_lazy_inplace"
+//
+// modifies its first argument
+// returns false if there wasn't enough memory for the operation
+// x1 will release its memory and be put in an invalid state on a false result
+//
+// Note from non-lazy form (is it true for lazy, too?)
+//    "XOR can have new containers inserted from x2, but can also
+//     lose containers when x1 and x2 are nonempty and identical."
+//
+static bool xor_inplace_core(
+    roaring_bitmap_t *x1,
+    const roaring_bitmap_t *x2,
+    bool (*dispatcher)(container_t**,uint8_t*,const container_t* c2,uint8_t)
+){
     assert(x1 != x2);
-    uint8_t result_type = 0;
     int length1 = x1->high_low_container.size;
     const int length2 = x2->high_low_container.size;
 
-    if (0 == length2) return;
+    if (0 == length2) return true;
 
     if (0 == length1) {
-        roaring_bitmap_overwrite(x1, x2);
-        return;
+        return roaring_bitmap_overwrite(x1, x2);
     }
     int pos1 = 0, pos2 = 0;
-    uint8_t type1, type2;
     uint16_t s1 = ra_get_key_at_index(&x1->high_low_container, pos1);
     uint16_t s2 = ra_get_key_at_index(&x2->high_low_container, pos2);
     while (true) {
         if (s1 == s2) {
-            container_t *c1 = ra_get_container_at_index(
-                                    &x1->high_low_container, pos1, &type1);
+            uint8_t *type1 = ra_typecode_slot(&x1->high_low_container, pos1);
+            container_t **c1 = ra_container_slot(&x1->high_low_container, pos1);
+
+            uint8_t type2;
             container_t *c2 = ra_get_container_at_index(
                                     &x2->high_low_container, pos2, &type2);
  
-            // We do the computation "in place" only when c1 is not a shared container.
-            // Rationale: using a shared container safely with in place computation would 
-            // require making a copy and then doing the computation in place which is likely 
-            // less efficient than avoiding in place entirely and always generating a new 
-            // container.
+            if (!(*dispatcher)(c1, type1, c2, type2))
+                goto alloc_failure;  // c1 should still be freeable
```

Comment:
I recommend using curly braces after an if clause *unless* the if clause fits on a single line.

Reply:
I never put a branch on the same line--because it can be extremely annoying when you want a breakpoint on the branch.  (While you can in some IDEs find a way to do it in the assembly view, the space savings is never worth having to do that.)  I also find it easier to overlook a branch on the same line.  The indentation calls out "hey, branch here!"

My opinion on braces for single-line conditions is that if modern compilers didn't have good warnings for catching unintended consequences, I'd certainly be for using them.  But todays warnings are good enough, so I find them to be visual noise.  This is not a huge deal to me.

*The main thing I think RE: formatting* is that it's a hassle whenever you require adjustments and alignments each time a name length changes.  Maybe it's appropriate for public-facing API headers that are going for a certain layout and look to be taken in more easily.  But for day-to-day code, I prefer:

     void some_function_name(
         int some_argument_name,
         int some_other_argument,
         int yet_another_argument
     ){
         /* function content here */
     }
     
...over more acrobatic forms which push things way off to the right, and require more maintenance and cognitive tax each time you want to make an adjustment.

Reply:
I am not very interested in style issues, but I am sure we agree that one should be reasonably consistent so that the whole code base has a somewhat uniform style. That's all I am saying.

I do not advocate anything more.